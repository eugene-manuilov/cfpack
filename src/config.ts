import { isAbsolute, resolve } from 'path';

import { RunnerContext, Stack } from './types';

/**
 * The class responsible for configuration managment.
 */
export class Config {

	/**
	 * Constructor.
	 *
	 * @since 2.0.0
	 *
	 * @param {RunnerContext} data The config data.
	 */
	public constructor( private readonly data: RunnerContext ) {}

	/**
	 * Loads a config file using initial configuration options provided using CLI args and returns a new config instance.
	 *
	 * @since 2.0.0
	 * @throws {Error} When config file can't be found.
	 *
	 * @static
	 * @access public
	 * @param {RunnerContext} data The initial configuration.
	 * @returns {Config} A new instance of the Config class.
	 */
	public static load( data: RunnerContext ): Config {
		try {
			const configPath = isAbsolute( data.config )
				? data.config
				: resolve( process.cwd(), data.config );

			const config = require( configPath );
			return new Config( { ...config, ...data } );
		} catch ( e ) {
			// do nothing.
		}

		throw new Error( 'Config file hasn\'t been found.' );
	}

	public get silent(): boolean {
		return !!this.data.silent;
	}

	public get verbose(): boolean {
		return !!this.data.verbose;
	}

	public get stack(): Stack {
		if ( this.data.stack ) {
			return this.data.stack;
		}

		return {} as Stack;
	}

	public get entry(): string {
		return this.data.entry || '';
	}

	public get output(): string {
		return this.data.output || '';
	}

	public get config(): string {
		return this.data.config;
	}

}
