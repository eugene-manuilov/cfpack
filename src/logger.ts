import { green, yellow, white, red } from 'chalk';
import * as ora from 'ora';

export class Logger {

	private readonly ora: ora.Ora = ora( {
		spinner: 'dots',
		color: 'white',
		hideCursor: true,
	} );

	public constructor(
		private readonly silent?: boolean,
		private readonly verbose?: boolean
	) {}

	public start(): void {
		if ( !this.silent ) {
			this.ora.start();
		}
	}

	public stop(): void {
		if ( this.ora.isSpinning ) {
			this.ora.stop();
		}
	}

	public sayIf( message: string, condition?: boolean ): void {
		if ( condition ) {
			this.stop();
			process.stdout.write( `${ message }\n` );
			this.start();
		}
	}

	public message( message: string ): void {
		this.sayIf( green( message ), !this.silent );
	}

	public info( message: string ): void {
		this.sayIf( white( message ), !this.silent && this.verbose );
	}

	public warning( message: string ): void {
		this.sayIf( yellow( message ), !this.silent && this.verbose );
	}

	public error( message: string, exit = true ): void {
		this.stop();
		process.stderr.write( red( `${ message }\n` ) );
		if ( exit ) {
			process.exit( 1 );
		}
	}

}
