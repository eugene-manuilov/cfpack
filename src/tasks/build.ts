import { tmpdir } from 'os';
import { isAbsolute, resolve, join, dirname } from 'path';
import {
	existsSync,
	readdirSync,
	statSync,
	readFileSync,
	mkdtempSync,
	writeFileSync
} from 'fs';

import { CloudFormation } from 'aws-sdk';
import { safeLoad, Schema } from 'js-yaml';
import { magenta } from 'chalk';

import { Task } from '../task';
import { createSchema } from '../utils';

export class BuildTask extends Task {

	private readonly intrinsicFunctionsSchema: Schema = createSchema();

	public run( next: Function ): void {
		if ( this.log ) {
			const { entry: entryFolder } = this.options || {};
			this.log.message( 'Build template file...' );
			this.log.info( `├─ Looking for templates in the ${ entryFolder } folder...` );
		}

		this.findTemplates();

		if ( !Array.isArray( this.output.files ) || this.output.files.length === 0 ) {
			return;
		}

		if ( this.log ) {
			this.log.info( '├─ Processing found templates...' );
		}

		this.processTemplates();

		if ( this.log ) {
			this.log.info( '├─ Validating final template...' );
		}

		this.validateFinalTemplate( () => {
			this.saveFinalTemplate();

			if ( this.log ) {
				this.log.info( `└─ Final template: ${ magenta( this.output.templateFile ) }\n` );
			}

			next( this.output );
		} );
	}

	public findTemplates(): void {
		const {
			entry = '',
			config = '',
		} = this.options || {};

		const entryPath = isAbsolute( entry )
			? entry
			: resolve( dirname( config ), entry );

		if ( !existsSync( entryPath ) && this.log ) {
			this.log.error( 'The entry folder is not found.' );
		}

		const files = this.walkTemplates( entryPath, [] );
		if ( this.log ) {
			if ( files.length > 0 ) {
				this.log.info( `├─ Found ${ files.length } template(s)...` );
			} else {
				this.log.info( '└─ Found no templates in the folder...' );
			}
		}

		this.output.files = files;
	}

	public walkTemplates( dir: string, list: string[] ): string[] {
		let newlist = [ ...list ];

		readdirSync( dir ).forEach( ( file ) => {
			const filename = join( dir, file );
			if ( statSync( filename ).isDirectory() ) {
				newlist = [ ...newlist, ...this.walkTemplates( filename, list ) ];
			} else {
				newlist.push( filename );
			}
		} );

		return newlist;
	}

	public processTemplates(): void {
		const template: { [x: string]: string | Date | {} } = {};
		const { files = [] } = this.output;

		files.forEach( ( file ) => {
			try {
				const doc = this.processTemplate( file );

				Object.keys( doc ).forEach( ( key: string ) => {
					const group: string | { [x: string]: unknown } | Date = doc[key];

					if ( group instanceof Date ) {
						const [ dateString ] = group.toISOString().split( 'T' );
						template[key] = dateString;
					} else if ( typeof group === 'object' ) {
						template[key] = { ...( template[key] || {} ), ...group };
					} else {
						template[key] = group;
					}
				} );

				if ( this.log ) {
					this.log.info( `├─ Processed ${ file } template...` );
				}
			} catch ( e ) {
				if ( this.log ) {
					this.log.info( `├─ Error processing ${ file } template: ${ e.toString().split( '\n' ).join( '\n│  ' ) }` );
				}
			}
		} );

		this.output.template = template;
	}

	public processTemplate( file: string ): { [x: string]: string | Date | {} } {
		const content = readFileSync( file, 'utf8' );
		return content.trim().charAt( 0 ) === '{'
			? JSON.parse( content )
			: safeLoad( content, { schema: this.intrinsicFunctionsSchema } );
	}

	public validateFinalTemplate( callback: Function ): void {
		const { stack } = this.options || {};
		const cloudformation = new CloudFormation( {
			apiVersion: '2010-05-15',
			region: stack ? stack.region : 'us-east-1',
		} );

		cloudformation.validateTemplate( { TemplateBody: JSON.stringify( this.output.template, undefined, 4 ) }, ( err ) => {
			if ( err ) {
				if ( this.log ) {
					this.log.error( `├─ ${ err.message }`, false );
					this.log.error( `└─ RequestId: ${ magenta( err.requestId ) }`, false );
					this.log.stop();
				}
				process.exit( 1 );
			} else {
				callback();
			}
		} );
	}

	public saveFinalTemplate(): void {
		let {
			config = '',
			output: filename,
		} = this.options || {};

		if ( !filename ) {
			const prefix = join( tmpdir(), 'cfpack-' );
			const folder = mkdtempSync( prefix );
			filename = join( folder, 'template.json' );
		}

		filename = isAbsolute( filename )
			? filename
			: resolve( dirname( config ), filename );

		const data = JSON.stringify( this.output.template, undefined, 4 );
		writeFileSync( filename, data, { encoding: 'utf8' } );
		this.output.templateFile = filename;
	}

}
