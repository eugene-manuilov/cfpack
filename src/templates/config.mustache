/**
 * The folder name where all CloudFormation templates are stored.
 *
 * @type {string}
 */
exports.entry = {{ entryFolder }};

/**
 * The filename for generated CloudFormation template.
 *
 * @type {string}
 */
exports.output = {{ outputFile }};

/**
 * Whether to display extended output information during commands run.
 *
 * @type {boolean}
 */
exports.verbose = true;

/**
 * Whether to suppress all output.
 *
 * @type {boolean}
 */
exports.silent = false;

/**
 * Stack declaration object.
 *
 * @type {object}
 */
exports.stack = {};

/**
 * The stack name.
 *
 * @type {string}
 */
exports.stack.name = {{ stackName }};

/**
 * The stack region name. It should be a valid region ID supported by AWS.
 *
 * @see https://aws.amazon.com/about-aws/global-infrastructure/#AWS_Global_Infrastructure_Map
 * @type {string}
 */
exports.stack.region = {{ stackRegion }};

/**
 * Additional parameters for the stack creation request.
 *
 * @see: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#createStack-property
 * @type {object}
 */
exports.stack.params = {
	{{ #capabilityIAM }}
	Capabilities: ['CAPABILITY_IAM'],
	{{ /capabilityIAM }}
	{{ ^capabilityIAM }}
	/* uncomment if your CloudFormation template manages IAM resources */
	// Capabilities: ['CAPABILITY_IAM'],
	{{ /capabilityIAM }}
	/* uncomment if your CloudFormation require parameters */
	// Parameters: [
	// 	{
	// 		ParameterKey: 'my-parameter',
	// 		ParameterValue: 'my-value',
	// 	},
	// ],
};

/**
 * Artifacts list to use for the stack. Artifacts are uploaded to a S3 bucket before stack is created or updated.
 *
 * @example
 * exports.stack.artifacts = [
 *   {
 *     bucket: 'my-s3-bucket',
 *     files: {
 *       'assets/images/': 'local/images/*.png'
 *       'assets/bundle.zip': {
 *         baseDir: 'local/files/',
 *         path: '*.js',
 *         compression: 'zip', // zip | none
 *       },
 *     },
 *   },
 * ];
 *
 * @type {Array}
 */
exports.stack.artifacts = [];
