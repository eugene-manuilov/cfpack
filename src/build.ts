import { tmpdir } from 'os';
import { isAbsolute, resolve, join, dirname } from 'path';
import {
	readFileSync,
	promises as fs
} from 'fs';

import { CloudFormation } from 'aws-sdk';
import { safeLoad, Schema } from 'js-yaml';
import { magenta } from 'chalk';

import { glob, createSchema } from './utils';
import { Logger } from './logger';
import { Config } from './config';
import { RunnerData } from './types';

export class BuildTask {

	private readonly intrinsicFunctionsSchema: Schema = createSchema();

	private files: string[] = [];

	private template: { [x: string]: {} } = {};

	private templateFile: string = '';

	public constructor(
		private readonly options: Config,
		private readonly log: Logger
	) {}

	public async run(): Promise<RunnerData> {
		const { entry: entryFolder = process.cwd(), config } = this.options;
		this.log.message( 'Build template file...' );
		this.log.info( `├─ Looking for templates in the ${ entryFolder } folder...` );

		const entryPath = isAbsolute( entryFolder )
			? entryFolder
			: resolve( dirname( config ), entryFolder );

		const entryPathStats = await fs.stat( entryPath ).catch( () => undefined );
		if ( !entryPathStats ) {
			this.log.error( 'The entry folder is not found.' );
			return {};
		}

		this.files = await glob( entryPath, '**/*.yaml', '**/*.yml', '**/*.json' );
		if ( this.files.length > 0 ) {
			this.log.info( `├─ Found ${ this.files.length } template(s)...` );
		} else {
			this.log.info( '└─ Found no templates in the folder...' );
			return {};
		}

		this.log.info( '├─ Processing found templates...' );
		this.processTemplates();

		this.log.info( '├─ Validating final template...' );
		await this.validateFinalTemplate();

		await this.saveFinalTemplate();
		this.log.info( `└─ Final template: ${ magenta( this.templateFile ) }\n` );

		return {
			template: this.template,
			templateFile: this.templateFile,
		};
	}

	public processTemplates(): void {
		const template: { [x: string]: string | Date | {} } = {};

		this.files.forEach( ( file ) => {
			try {
				const doc = this.processTemplate( file );

				Object.keys( doc ).forEach( ( key: string ) => {
					const group: string | { [x: string]: unknown } | Date = doc[key];

					if ( group instanceof Date ) {
						const [ dateString ] = group.toISOString().split( 'T' );
						template[key] = dateString;
					} else if ( typeof group === 'object' ) {
						template[key] = { ...( template[key] || {} ), ...group };
					} else {
						template[key] = group;
					}
				} );

				this.log.info( `├─ Processed ${ file } template...` );
			} catch ( e ) {
				this.log.info( `├─ Error processing ${ file } template: ${ e.toString().split( '\n' ).join( '\n│  ' ) }` );
			}
		} );

		this.template = template;
	}

	public processTemplate( file: string ): { [x: string]: string | Date | {} } {
		const content = readFileSync( file, { encoding: 'utf8' } );
		return content.trim().charAt( 0 ) === '{'
			? JSON.parse( content )
			: safeLoad( content, { schema: this.intrinsicFunctionsSchema } );
	}

	public async validateFinalTemplate(): Promise<void> {
		try {
			const { stack } = this.options;
			const body = JSON.stringify( this.template, undefined, 4 );
			const cloudformation = new CloudFormation( {
				apiVersion: '2010-05-15',
				region: stack ? stack.region : 'us-east-1',
			} );

			await cloudformation.validateTemplate( { TemplateBody: body } ).promise();
		} catch ( err ) {
			this.log.error( `├─ ${ err.message }`, false );
			this.log.error( `└─ RequestId: ${ magenta( err.requestId ) }`, false );
			this.log.stop();
			process.exit( 1 );
		}
	}

	public async saveFinalTemplate(): Promise<void> {
		let {
			config = '',
			output: filename,
		} = this.options || {};

		if ( !filename ) {
			const prefix = join( tmpdir(), 'cfpack-' );
			const folder = await fs.mkdtemp( prefix );
			filename = join( folder, 'template.json' );
		}

		this.templateFile = isAbsolute( filename )
			? filename
			: resolve( dirname( config ), filename );

		await fs.writeFile(
			this.templateFile,
			JSON.stringify( this.template, undefined, 4 ),
			{ encoding: 'utf8' }
		);
	}

}
