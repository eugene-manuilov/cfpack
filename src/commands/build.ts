import { Command, flags } from '@oclif/command';

import { BuildTask } from '../tasks/build';
import { Config } from '../config';
import { Logger } from '../logger';

class BuildCommand extends Command {

	public static description = 'Assembles templates into one CloudFormation template.';

	public static flags = {
		config: flags.string( {
			description: 'Path to the config file.',
			default: 'cfpack.config.js',
			required: false,
		} ),
		verbose: flags.boolean( {
			description: 'Display extended details.',
			required: false,
		} ),
		silent: flags.boolean( {
			description: 'Prevent output from being displayed in stdout.',
			required: false,
		} ),
	};

	public run(): Promise<void> {
		return new Promise( ( resolve ) => {
			const { flags } = this.parse( BuildCommand );
			const config = Config.load( flags );
			const logger = new Logger( config.silent, config.verbose );

			const builder = new BuildTask();
			builder.setLogger( logger );
			builder.setOptions( config );
			builder.run( resolve );
		} );
	}

}

export default BuildCommand;
