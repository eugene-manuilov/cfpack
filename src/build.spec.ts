import { resolve } from 'path';

import { BuildTask } from './build';
import { Config } from './config';
import { Logger } from './logger';

describe( 'BuildTask', () => {
	test.each( [
		'and.yml',
		'equals.yml',
		'if.yml',
		'not.yml',
		'or.yml',
		'base64.yml',
		'cidr.yml',
		'find-in-map.yml',
		'get-att.yml',
		'get-azs.yml',
		'import-value.yml',
		'join.yml',
		'select.yml',
		'split.yml',
		'sub.yml',
		'transform.yml',
		'ref.yml',
		// @todo: add template to test !Sub function
	] )( '::processTemplate --> {%s}', ( template ) => {
		const config = new Config( {
			config: '',
			silent: true,
			verbose: false,
		} );

		const task = new BuildTask( config, new Logger() );
		const filename = resolve( __dirname, '__fixtures__', template );

		let result;

		expect( () => { result = task.processTemplate( filename ); } ).not.toThrow();
		expect( JSON.stringify( result, undefined, 2 ) ).toMatchSnapshot();
	} );
} );
