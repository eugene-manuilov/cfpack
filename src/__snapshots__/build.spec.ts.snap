// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildTask ::processTemplate --> {and.yml} 1`] = `
"{
  \\"MyAndCondition\\": {
    \\"Fn::And\\": [
      {
        \\"Fn::Equals\\": [
          \\"sg-mysggroup\\",
          {
            \\"Ref\\": \\"ASecurityGroup\\"
          }
        ]
      },
      {
        \\"Condition\\": \\"SomeOtherCondition\\"
      }
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {base64.yml} 1`] = `
"{
  \\"Fn::Base64\\": \\"AWS CloudFormation\\"
}"
`;

exports[`BuildTask ::processTemplate --> {cidr.yml} 1`] = `
"{
  \\"Resources\\": {
    \\"ExampleVpc\\": {
      \\"Type\\": \\"AWS::EC2::VPC\\",
      \\"Properties\\": {
        \\"CidrBlock\\": \\"10.0.0.0/16\\"
      }
    },
    \\"IPv6CidrBlock\\": {
      \\"Type\\": \\"AWS::EC2::VPCCidrBlock\\",
      \\"Properties\\": {
        \\"AmazonProvidedIpv6CidrBlock\\": true,
        \\"VpcId\\": {
          \\"Ref\\": \\"ExampleVpc\\"
        }
      }
    },
    \\"ExampleSubnet\\": {
      \\"Type\\": \\"AWS::EC2::Subnet\\",
      \\"DependsOn\\": \\"IPv6CidrBlock\\",
      \\"Properties\\": {
        \\"AssignIpv6AddressOnCreation\\": true,
        \\"CidrBlock\\": {
          \\"Fn::Select\\": [
            0,
            {
              \\"Fn::Cidr\\": [
                {
                  \\"Fn::GetAtt\\": [
                    \\"ExampleVpc\\",
                    \\"CidrBlock\\"
                  ]
                },
                1,
                8
              ]
            }
          ]
        },
        \\"Ipv6CidrBlock\\": {
          \\"Fn::Select\\": [
            0,
            {
              \\"Fn::Cidr\\": [
                {
                  \\"Fn::Select\\": [
                    0,
                    {
                      \\"Fn::GetAtt\\": [
                        \\"ExampleVpc\\",
                        \\"Ipv6CidrBlocks\\"
                      ]
                    }
                  ]
                },
                1,
                64
              ]
            }
          ]
        },
        \\"VpcId\\": {
          \\"Ref\\": \\"ExampleVpc\\"
        }
      }
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {equals.yml} 1`] = `
"{
  \\"UseProdCondition\\": {
    \\"Fn::Equals\\": [
      {
        \\"Ref\\": \\"EnvironmentType\\"
      },
      \\"prod\\"
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {find-in-map.yml} 1`] = `
"{
  \\"Mappings\\": {
    \\"RegionMap\\": {
      \\"us-east-1\\": {
        \\"HVM64\\": \\"ami-0ff8a91507f77f867\\",
        \\"HVMG2\\": \\"ami-0a584ac55a7631c0c\\"
      },
      \\"us-west-1\\": {
        \\"HVM64\\": \\"ami-0bdb828fd58c52235\\",
        \\"HVMG2\\": \\"ami-066ee5fd4a9ef77f1\\"
      },
      \\"eu-west-1\\": {
        \\"HVM64\\": \\"ami-047bb4163c506cd98\\",
        \\"HVMG2\\": \\"ami-31c2f645\\"
      },
      \\"ap-southeast-1\\": {
        \\"HVM64\\": \\"ami-08569b978cc4dfa10\\",
        \\"HVMG2\\": \\"ami-0be9df32ae9f92309\\"
      },
      \\"ap-northeast-1\\": {
        \\"HVM64\\": \\"ami-06cd52961ce9f0d85\\",
        \\"HVMG2\\": \\"ami-053cdd503598e4a9d\\"
      }
    }
  },
  \\"Resources\\": {
    \\"myEC2Instance\\": {
      \\"Type\\": \\"AWS::EC2::Instance\\",
      \\"Properties\\": {
        \\"ImageId\\": {
          \\"Fn::FindInMap\\": [
            \\"RegionMap\\",
            {
              \\"Ref\\": \\"AWS::Region\\"
            },
            \\"HVM64\\"
          ]
        },
        \\"InstanceType\\": \\"m1.small\\"
      }
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {get-att.yml} 1`] = `
"{
  \\"AWSTemplateFormatVersion\\": \\"2010-09-09T00:00:00.000Z\\",
  \\"Resources\\": {
    \\"myELB\\": {
      \\"Type\\": \\"AWS::ElasticLoadBalancing::LoadBalancer\\",
      \\"Properties\\": {
        \\"AvailabilityZones\\": [
          \\"eu-west-1a\\"
        ],
        \\"Listeners\\": [
          {
            \\"LoadBalancerPort\\": \\"80\\",
            \\"InstancePort\\": \\"80\\",
            \\"Protocol\\": \\"HTTP\\"
          }
        ]
      }
    },
    \\"myELBIngressGroup\\": {
      \\"Type\\": \\"AWS::EC2::SecurityGroup\\",
      \\"Properties\\": {
        \\"GroupDescription\\": \\"ELB ingress group\\",
        \\"SecurityGroupIngress\\": [
          {
            \\"IpProtocol\\": \\"tcp\\",
            \\"FromPort\\": \\"80\\",
            \\"ToPort\\": \\"80\\",
            \\"SourceSecurityGroupOwnerId\\": {
              \\"Fn::GetAtt\\": [
                \\"myELB\\",
                \\"SourceSecurityGroup.OwnerAlias\\"
              ]
            },
            \\"SourceSecurityGroupName\\": {
              \\"Fn::GetAtt\\": [
                \\"myELB\\",
                \\"SourceSecurityGroup.GroupName\\"
              ]
            }
          }
        ]
      }
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {get-azs.yml} 1`] = `
"{
  \\"evaluateRegions\\": [
    {
      \\"Fn::GetAZs\\": \\"\\"
    },
    {
      \\"Fn::GetAZs\\": {
        \\"Ref\\": \\"AWS::Region\\"
      }
    },
    {
      \\"Fn::GetAZs\\": \\"us-east-1\\"
    }
  ],
  \\"mySubnet\\": {
    \\"Type\\": \\"AWS::EC2::Subnet\\",
    \\"Properties\\": {
      \\"VpcId\\": {
        \\"Ref\\": \\"VPC\\"
      },
      \\"CidrBlock\\": \\"10.0.0.0/24\\",
      \\"AvailabilityZone\\": {
        \\"Fn::Select\\": [
          0,
          {
            \\"Fn::GetAZs\\": \\"\\"
          }
        ]
      }
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {if.yml} 1`] = `
"{
  \\"SecurityGroups\\": [
    {
      \\"Fn::If\\": [
        \\"CreateNewSecurityGroup\\",
        {
          \\"Ref\\": \\"NewSecurityGroup\\"
        },
        {
          \\"Ref\\": \\"ExistingSecurityGroup\\"
        }
      ]
    }
  ],
  \\"Outputs\\": {
    \\"SecurityGroupId\\": {
      \\"Description\\": \\"Group ID of the security group used.\\",
      \\"Value\\": {
        \\"Fn::If\\": [
          \\"CreateNewSecurityGroup\\",
          {
            \\"Ref\\": \\"NewSecurityGroup\\"
          },
          {
            \\"Ref\\": \\"ExistingSecurityGroup\\"
          }
        ]
      }
    }
  },
  \\"MyDB\\": {
    \\"Type\\": \\"AWS::RDS::DBInstance\\",
    \\"Properties\\": {
      \\"AllocatedStorage\\": 5,
      \\"DBInstanceClass\\": \\"db.m1.small\\",
      \\"Engine\\": \\"MySQL\\",
      \\"EngineVersion\\": 5.5,
      \\"MasterUsername\\": {
        \\"Ref\\": \\"DBUser\\"
      },
      \\"MasterUserPassword\\": {
        \\"Ref\\": \\"DBPassword\\"
      },
      \\"DBParameterGroupName\\": {
        \\"Ref\\": \\"MyRDSParamGroup\\"
      },
      \\"DBSnapshotIdentifier\\": {
        \\"Fn::If\\": [
          \\"UseDBSnapshot\\",
          {
            \\"Ref\\": \\"DBSnapshotName\\"
          },
          {
            \\"Ref\\": \\"AWS::NoValue\\"
          }
        ]
      }
    }
  },
  \\"UpdatePolicy\\": {
    \\"AutoScalingRollingUpdate\\": {
      \\"Fn::If\\": [
        \\"RollingUpdates\\",
        {
          \\"MaxBatchSize\\": 2,
          \\"MinInstancesInService\\": 2,
          \\"PauseTime\\": \\"PT0M30S\\"
        },
        {
          \\"Ref\\": \\"AWS::NoValue\\"
        }
      ]
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {import-value.yml} 1`] = `
"{
  \\"Resources\\": {
    \\"Service\\": {
      \\"Type\\": \\"AWS::ECS::Service\\",
      \\"Properties\\": {
        \\"Cluster\\": {
          \\"Fn::ImportValue\\": \\"ClusterName\\"
        }
      }
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {join.yml} 1`] = `
"{
  \\"JoinWithParams\\": {
    \\"Fn::Join\\": [
      \\"\\",
      [
        \\"arn:\\",
        {
          \\"Ref\\": \\"Partition\\"
        },
        \\":s3:::elasticbeanstalk-*-\\",
        {
          \\"Ref\\": \\"AWS::AccountId\\"
        }
      ]
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {not.yml} 1`] = `
"{
  \\"MyNotCondition\\": {
    \\"Fn::Not\\": [
      {
        \\"Fn::Equals\\": [
          {
            \\"Ref\\": \\"EnvironmentType\\"
          },
          \\"prod\\"
        ]
      }
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {or.yml} 1`] = `
"{
  \\"MyOrCondition\\": {
    \\"Fn::Or\\": [
      {
        \\"Fn::Equals\\": [
          \\"sg-mysggroup\\",
          {
            \\"Ref\\": \\"ASecurityGroup\\"
          }
        ]
      },
      {
        \\"Condition\\": \\"SomeOtherCondition\\"
      }
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {ref.yml} 1`] = `
"{
  \\"MyEIP\\": {
    \\"Type\\": \\"AWS::EC2::EIP\\",
    \\"Properties\\": {
      \\"InstanceId\\": {
        \\"Ref\\": \\"MyEC2Instance\\"
      }
    }
  }
}"
`;

exports[`BuildTask ::processTemplate --> {select.yml} 1`] = `
"{
  \\"Basic\\": {
    \\"Fn::Select\\": [
      \\"1\\",
      [
        \\"apples\\",
        \\"grapes\\",
        \\"oranges\\",
        \\"mangoes\\"
      ]
    ]
  },
  \\"Subnet0\\": {
    \\"Type\\": \\"AWS::EC2::Subnet\\",
    \\"Properties\\": {
      \\"VpcId\\": {
        \\"Ref\\": \\"VPC\\"
      },
      \\"CidrBlock\\": {
        \\"Fn::Select\\": [
          0,
          {
            \\"Ref\\": \\"DbSubnetIpBlocks\\"
          }
        ]
      }
    }
  },
  \\"AvailabilityZone\\": {
    \\"Fn::Select\\": [
      0,
      {
        \\"Fn::GetAZs\\": {
          \\"Ref\\": \\"AWS::Region\\"
        }
      }
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {split.yml} 1`] = `
"{
  \\"SimpleList\\": {
    \\"Fn::Split\\": [
      \\"|\\",
      \\"a|b|c\\"
    ]
  },
  \\"WithEmptyValues\\": {
    \\"Fn::Split\\": [
      \\"|\\",
      \\"a||c|\\"
    ]
  },
  \\"ImportedOutputValue\\": {
    \\"Fn::Select\\": [
      2,
      {
        \\"Fn::Split\\": [
          \\",\\",
          {
            \\"Fn::ImportValue\\": \\"AccountSubnetIDs\\"
          }
        ]
      }
    ]
  }
}"
`;

exports[`BuildTask ::processTemplate --> {transform.yml} 1`] = `
"{
  \\"TransformTest\\": {
    \\"Fn::Transform\\": {
      \\"Name\\": \\"AWS::Include\\",
      \\"Parameters\\": {
        \\"Location\\": \\"us-east-1\\"
      }
    }
  }
}"
`;
